### PR TITLE
Update the logic for local attetion mask in ModernBERT

### PR DESCRIPTION
**Pull request by kozistr**
# What does this PR do?

In the previous PR #459, `local_attention` can only be worked on `CPU` [due to the `abs()` operation](https://github.com/huggingface/text-embeddings-inference/pull/459#issuecomment-2779101344).

So, I've made a change to calculate the `window mask` in pure Rust and then create the Tensor from it. Maybe, this could make `ModernBERT` run on both MPS and CUDA too.

I've checked the output with [this script](https://github.com/huggingface/text-embeddings-inference/pull/459#issuecomment-2764840553) and it seems identical to before.

- [x] tested on CPU (local machine, WSL2)
- [x] tested on T4 GPU (Kaggle notebook, TEI server runs successfully)
- [ ] tested on MPS (might work, but needs to be tested)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@Narsil, @alvarobartt, @ivarflakstad

### CODE REVIEW

File: backends/candle/src/models/modernbert.rs [hunk 1]
```diff
@@ -577,28 +577,49 @@ impl ModernBertModel {
...
-            .unsqueeze(0)?
-            .unsqueeze(0)?
-            .broadcast_as(mask_shape)?;
+        let seq_len = global_attention_mask.dim(2)?;
+        let window_mask = self.get_window_mask(seq_len)?;
+        let window_mask = (window_mask * min_value)?.to_dtype(self.dtype)?;
 
-        let zero_tensor = Tensor::zeros_like(&attention_mask)?;
-        let local_attention_mask = attention_mask.where_cond(&window_mask, &zero_tensor)?;
+        let local_attention_mask = global_attention_mask.broadcast_add(&window_mask)?;
```
Comment (kozistr) [inline backends/candle/src/models/modernbert.rs:620]: I just added them instead of using masked_fill (or where_cond), which is referred to [here](https://github.com/huggingface/candle/blob/main/candle-transformers/src/models/modernbert.rs#L226).

When values from both masks (`global_attention_mask` and `window_mask`) are `f32::MIN`, the output becomes f32::MIN + f32::MIN, and it could overflow. I honestly haven't looked into it in detail, so we need to check that it's okay.

**Comment by Narsil** [comment]
@kozistr I think everything already works on MPS, it's been checked by @ErikKaum https://github.com/huggingface/text-embeddings-inference/pull/562

During the flash fix I forced the mask creation in CPU candle (to enable non-flash CUDA execution), and that fixed metal as well.
I think keeping everything in candle instead of Rust, makes it slightly nicer, because we could switch to on-device execution if we simply created the missing kernels (which is relatively easy to do).

No matter what, having a mask that is only created in the first part of the model is acceptable imho.

**Comment by Narsil** [comment]
That being said, we could definitely clean up a little bit that code imho.

The HashMap<bool, XX> could be replaced with a simple `struct Mask{local: Tensor, global: Tensor}` which should increase readability (and the overhead of a HashMap is not negligible even though I doubt it can be measure in this particular instance).

**Comment by kozistr** [comment]
> @kozistr I think everything already works on MPS, it's been checked by @ErikKaum #562
> 
> During the flash fix I forced the mask creation in CPU candle (to enable non-flash CUDA execution), and that fixed metal as well. I think keeping everything in candle instead of Rust, makes it slightly nicer, because we could switch to on-device execution if we simply created the missing kernels (which is relatively easy to do).
> 
> No matter what, having a mask that is only created in the first part of the model is acceptable IMHO.

I thought there were still some issues on MPS, but I missed that PR. Thanks for checking on that.
And I agree with your point that keeping everything in Candle would be great!

> That being said, we could definitely clean up a little bit that code imho.
> 
> The HashMap<bool, XX> could be replaced with a simple `struct Mask{local: Tensor, global: Tensor}` which should increase readability (and the overhead of a HashMap is not negligible even though I doubt it can be measure in this particular instance).

I second this. We could refactor the mask part with `struct` instead of `HashMap` regarding readability.


Then, I'm gonna close this PR and make another contribution later!
 
Thanks for your time to review this PR :)

**Comment by Narsil** [comment]
Thanks a lot to you.
